src/hotspot/share/runtime/javaThread.cpp

```diff
@@ -1015,6 +1016,36 @@ void JavaThread::set_exception_oop(oop o) {
   Atomic::store(&_exception_oop, o);
 }
 
+// Effectively wait for any deoptimization to complete and clear _monitor_chunk.
+class ReadMonitorChunksHandshake : public HandshakeClosure {
+  MonitorChunk* _mc;
+ public:
+  ReadMonitorChunksHandshake(): HandshakeClosure("ReadMonitorChunks") {}
+  void do_thread(Thread* thr) {
+    JavaThread* target = JavaThread::cast(thr);
+    _mc = target->monitor_chunks();
+  }
+  MonitorChunk* monitor_chunks() {
+    return _mc;
+  }
+};
+
+// _monitor_chunks is set and cleared by deoptimization.
+// If non-null, deoptimization is happening and the value
+// will be cleared soon.  Waiting for deopt is the only action.
+// A ThreadLocalHandshake will mean deopt is complete.
+MonitorChunk* JavaThread::monitor_chunks_safe() const {
+  MonitorChunk* chunks = _monitor_chunks;
```

Comment:
There's still a race here, right?  The target thread isn't necessarily the same as the current thread, so it could deoptimize immediately after reading `_monitor_chunks`.  I think for correctness it would always need to handshake.  But then what's the point, if we always return null?  Then this just turns into a strange synchronization mechanism.

Reply:
Right, target thread is commonly not the current thread (hence the occasional problem).

At the moment yes what I have here is a strange synchronziation, waiting for a thing to be null, then returning null.
It's a step towards removing the callers of monitor_chunks() from the few places that are not deoptimization itself.  Only waiting really to do the assert.  As you notice the call from rootResolver.cpp looks safe to remove: I see it is called from a safepoint.

In this one in javaThread.cpp, yes it could deoptimize and set _monitor_chunks after we read it.  If it does that...  I'm checking into what can happen...